Added a simple version of a polling loop.

kodek16 opened the pull request:
`dune build -w` now behaves in a similar way to Jenga polling mode,
automatically rebuilding the project when changes are detected.
`dune runtest -w` is similar, but also runs tests after successful
build.

Note that on platforms other than Linux, you need to have `fswatch`
installed for watch mode to work.

diml commented on bin/main.ml:
```diff
@@ -749,6 +782,47 @@ let resolve_targets_exn ~log common setup user_targets =
       targets)
 
+let fswatch_command root_path =
+  let excludes = [ {|/\.#|}
```
A few elements in this list only seem relevant when not doing out-of-source builds. Let's restrict it to:
```
                  [ {|\.#|}
                  ; {|_build|}
￼                  ; {|\.hg|}
￼                  ; {|\.git|}
￼                  ; {|~$|}
￼                  ; {|/#[^#]*#$|}
￼                  ; {|\.install$|}
                  ]
```

diml commented on bin/main.ml:
```diff
@@ -769,10 +843,67 @@ let build_targets =
     let log = Log.create common in
-    Scheduler.go ~log ~common
-      (Main.setup ~log common >>= fun setup ->
+
```
The polling mode is useful for both `dune build` and `dune runtest`. Could you factorize the code so that it works with both commands?

diml commented on bin/main.ml:
```diff
@@ -769,10 +843,67 @@ let build_targets =
+         >>= fun () ->
+         fswatch_command File_tree.(Dir.path (root setup.file_tree))
+         >>= fun (fswatch, args) ->
+         let fswatch =
```
We should do this outside of the loop. `Dir.path (root setup.file_tree)` is always `Path.root`

diml commented:
One thing I just remember we need to handle is promotion. Sometimes, dune records  files that needs to be promoted to the source tree. When `--auto-promote` is passed, these files should be promoted at the end of the build and we should loop immediately. The `Promotion` module records the files that need to be promoted

Khady commented:
Does it detect changes during the compilation? I had a quick look at the code, but it's not obvious to me. As soon as the compilation takes more than a few seconds, this is a feature I feel is required.

kodek16 commented:
@Khady, file changes while build is active are not detected yet, but I agree this is definitely something to be improved in near future.

kodek16 commented:
@diml, auto-rebuilds after promotions will be solved in a natural way once build restart on changes is implemented. For now I've added a simple hackish flag to retrigger build if at least one file was promoted.

emillon commented on bin/main.ml:
```diff
@@ -749,6 +788,56 @@ let resolve_targets_exn ~log common setup user_targets =
+  in
+  let path = Path.to_string_maybe_quoted root_path in
+  let uname =
+    match Bin.which "uname" with
```
Rather than doing this in two steps (and relying on `uname`), can we first check if `inotifywait` exists and use it if it does?

kodek16 replied:
Good point.

emillon commented on bin/main.ml:
```diff
@@ -749,6 +788,56 @@ let resolve_targets_exn ~log common setup user_targets =
+      Format.eprintf "Error: executable not found: %s\n" fswatch;
+      exit 1
+  in
+  Process.run_capture Strict fswatch args ~env:Env.initial
```
I'm not sure that this is intended, but it seems that all call sites of `watch_changes` pipe it into `ignore`, so it might be worth returning `unit` directly instead.

diml replied:
Or even: `Process.run ~stdout:Config.dev_null`

kodek16 replied:
I've added `>>| ignore` for now, but in the future we're likely to need the actual names of files that were modified, so this is going to change anyway.

diml commented on bin/main.ml:
```diff
@@ -139,6 +142,35 @@ module Scheduler = struct
+      >>= fun _ ->
+      main_loop ()
+    in
+    try
```
We can simplify this code with a single loop: `let rec loop f = try Scheduler.go ... (f ()) ...`

diml commented on bin/main.ml:
```diff
@@ -769,10 +850,35 @@ let build_targets =
+      let targets = resolve_targets_exn ~log common setup targets in
+      do_build setup targets
+    in
+    if common.watch then
```
This code looks very similar to the one for `dune runtest`. It can be factorized a bit more

kodek16 replied:
Done.

diml commented:
I tried it again and it seems to work well when combined with `--auto-promote`. However, the messages `Success, polling ...` are now gone.

diml commented:
Never mind, I tried again and it works. I was using `dune exec -- dune runtest -w`. This must be a bug in `dune exec`

kodek16 commented:
One other thing that we should consider is the naming: do you think we should stick with `-w` for "watch mode", or switch to `-p` for polling mode as in Jenga?

diml commented on bin/main.ml:
```diff
@@ -130,6 +133,43 @@ module Log = struct
+       fswatch, ["-r"; path; "-1"] @ excludes
+     | None ->
+      (* Exit immediately to prevent a loop of these errors. *)
+      Format.eprintf "Error: fswatch not found. \
```
To match the style of other errors, could you use:

```ocaml
die "@{<error>Error@}: fswatch ..."
```

The message needs to mention inotify as well since we look for either of these tools

diml commented on src/scheduler.mli:
```diff
@@ -2,21 +2,45 @@
 
 open! Stdune
 
+type status_line_config = string option * [`Show_jobs | `Don't_show_jobs]
```
This type seems simpler to me:

```ocaml
type status_line_config =
  { message : string option
  ; show_jobs : bool
  }
```

diml commented on src/scheduler.ml:
```diff
@@ -190,18 +195,109 @@ let go ?(log=Log.no_log) ?(config=Config.default)
+      Format.eprintf "Internal error: setup failed.\n";
+      exit 1
+  in
+  let init =
```
This should be `init ()`. In general we assume that a given `Fiber.t` value is executed only once

diml commented:
Could you try adding a module Hooks with the following signature:

```ocaml
(** Register a function called at the end of every build *)
val at_end_of_build : (unit -> unit) -> unit

(** Regsiter a function called at the end of the current built only *)
val at_end_of_current_build : (unit -> unit) -> unit

(** Invoke all registered hooks *)
val end_of_build : (unit -> unit)
```

and use it to deal with the various cleanup functions? I feel like this would simplify a bit the code

diml commented:
> One other thing that we should consider is the naming: do you think we should stick with -w for "watch mode", or switch to -p for polling mode as in Jenga?

Unfortunately, we are already using `-p` for release builds. I.e. the build command packages must use when built with opam is: `dune build -p <package>`.

diml commented on bin/main.ml:
```diff
@@ -130,6 +133,41 @@ module Log = struct
+                      One of them needs to be installed for watch mode to work.\n")
+
+let watch_changes () =
+  let watch, args = watch_command Path.root in
```
We should cache the result of this call with a lazy

diml commented on bin/main.ml:
```diff
@@ -753,6 +814,33 @@ let resolve_targets_exn ~log common setup user_targets =
     | Ok targets ->
       targets)
 
+let clear_all_caches () =
```
Can't we put these in hooks?

diml commented on src/scheduler.mli:
```diff
@@ -2,21 +2,48 @@
+  :  ?log:Log.t
+  -> ?config:Config.t
+  -> ?cache_init:bool
+  -> init:(unit -> 'a Fiber.t)
```
I believe we can now change this to `unit -> unit Fiber.t`, which should simplify a bit the implementation. It also seems to me that `once` and `finally` should have type `... -> unit Fiber.t`, given that we currently do nothing with `'b` or `'c`.

diml commented:
@kodek16, it's possible that we'll need to provide more granularity in hooks. For instance to allows to recognize the end of the a build vs an interruption of the current build. For long running builds, it is also likely that we will need to add a mechanism to dump the databases at regular intervals. The hooks interface allows to keep all the logic related to a system in its own module, which is nice.

BTW, We should put `End_of_build.run` in an `at_exit`, to make sure it's always called.

emillon commented:
I tried this and it looks like it will be very useful! I have some remarks:
- when `fswatch` and `inotifywait` are not installed, this outputs errors in an infinite loop, taking 100% CPU.
- when running `dune build -w`, after fixing an error, the only feedback that the error is fixed is the "Success" message that is on the same line as the "polling" message. May I suggest clearing the terminal (using `Bin.which "clear"` if available), or at least displaying a message on its own line? The same happens if we go from a success state to a success state: there's little feedback that the build has completed.
- "polling filesystem" in the message makes it look like we're actively polling the filesystem rather than subscribed to events (some watch tools have a fallback mode like that). What do you think about a more neutral "waiting for filesystem changes"?
- can you rebase this on top of `master`? the history is a bit messy at the moment.

And once again thanks for working on this! :sparkles:

kodek16 committed:
Squashed commit of the following:

commit ca3c1b7a3e9bd8cec1788228adda6e6d46fa610f
Merge: f1744b6d 610b20d8
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Tue Sep 4 10:28:04 2018 -0400

    Merge branch 'master' into watch-mode

commit f1744b6dbdc142f51c08ab08d85553dee447be45
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 31 16:13:53 2018 -0400

    Add Hooks module.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 09e0de6a442e7f74e8bc02cfec59b738027bf8c0
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 31 15:08:28 2018 -0400

    Fix review suggestions.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit a59d70dfe74a23a01c7a398f04f54607cb322a8d
Merge: c7aa8ed8 6da1f199
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 31 14:35:50 2018 -0400

    Merge branch 'master' into watch-mode

commit c7aa8ed8587a0599caf0584f69f64fc5db97ac3f
Merge: aa15c6cb 6dedc681
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 31 09:37:28 2018 -0400

    Merge branch 'master' into watch-mode

commit aa15c6cbc4f12478b91826337f92dade7c6bc4b5
Merge: ce562e43 265adc8e
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Wed Aug 29 11:04:25 2018 -0400

    Merge branch 'master' into watch-mode

commit ce562e433ac9e43f50e071c2bdc2aaec217d6678
Merge: e8b55b4e c0eebce1
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Mon Aug 27 15:05:51 2018 -0400

    Merge branch 'master' into watch-mode

commit e8b55b4e09678d70cb42fa674baaa1dbca8e601c
Merge: 1a1289fb 9d61c595
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Mon Aug 27 11:31:54 2018 -0400

    Merge branch 'master' into watch-mode

commit 1a1289fbb6ec13b706a041710a2513afbd51cb24
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 24 10:40:19 2018 +0100

    Fixed build after merge.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit c2288e091fee2d929d742e36906802f2eebe98c7
Merge: d5fd1c4e b7da0304
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 24 10:26:27 2018 +0100

    Merge branch 'master' into watch-mode

commit d5fd1c4ee798b855e22c571327f8538dd1ecbc51
Merge: d69b6c5f b424ba97
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Tue Aug 21 11:38:07 2018 +0100

    Merge branch 'master' into watch-mode

commit d69b6c5f053c03da8e43e7fa042de777271fcc8e
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Tue Aug 21 11:26:21 2018 +0100

    Moved polling logic inside Scheduler module.

    This prevents us from calling Scheduler.go multiple times
    which may be problematic because of scheduler state.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit f76d3d19d4f37814b7f2be2929d874d246c232fa
Merge: e4935fbd 5b6496fa
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Tue Aug 21 09:45:58 2018 +0100

    Merge branch 'master' into watch-mode

commit e4935fbd1317e66f005621f7ded531eab4b18f08
Merge: 584b9e0a 6045568e
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Mon Aug 20 17:05:14 2018 +0100

    Merge branch 'master' into watch-mode

commit 584b9e0a0e9c6b3f7166b401770e707173c5ed2e
Merge: 6740ec8a 99d82c23
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Mon Aug 20 09:51:10 2018 +0100

    Merge branch 'master' into watch-mode

commit 6740ec8a69222425d73c68928190bade8883cfdd
Merge: a3cbc56b 309dd5ff
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 17 15:21:22 2018 +0100

    Merge branch 'master' into watch-mode

commit a3cbc56b5d970254529fe1def084d1470f220112
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 17 14:59:19 2018 +0100

    Fixed non-polling build.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 882c5d526d410478f809270e7961cf9986635c09
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Fri Aug 17 12:59:13 2018 +0100

    Refactored polling to only call finalize once.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 3e4f17a642b9d06897bba0fedbfdafc1b0f136b6
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Thu Aug 16 15:27:51 2018 +0100

    Made status line nicer and fixed auto-promote polling.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit f84cda3a48e2afa4e64a294d43d9c0def9befb29
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Thu Aug 16 14:59:50 2018 +0100

    Updated tests, removed duplicate finalize call.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 88a13731dfae6abcaa333659615c092ffedd729d
Merge: 8f05ef5f 166dd068
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Thu Aug 16 14:31:04 2018 +0100

    Merge branch 'master' into watch-mode

commit 8f05ef5f7ee4e53edc0e46f11ac758e47ae62bf3
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Thu Aug 16 14:27:43 2018 +0100

    Fixed promotion behavior.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit b01f6031856ef4338794bc6d6ef8dd7f9dafe919
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Thu Aug 16 10:23:38 2018 +0100

    Refactored.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 40f63cb812a8ee31b3b5d0606ae8f59de0e5ce68
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Thu Aug 16 09:51:16 2018 +0100

    Removed call to uname.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit ac2aaa79a357c496f0b11fd0562d2368c91e8be2
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Wed Aug 15 16:47:03 2018 +0100

    Made polling mode auto-rebuild after promotions.

    This is a temporary solution until automatic build restarts
    are implemented.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 55fc6990e2db23fcf306c9c73bb7545d08a09827
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Wed Aug 15 16:10:53 2018 +0100

    Refactored polling to be more abstract.

    Polling now has better scheduler API, which allows to use
    it in `dune runtest` without duplicating too much code.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit 1146d42f2ea8daef8a4d089841956600f4fcf5d2
Author: Jeremie Dimino <jeremie@dimino.org>
Date:   Wed Aug 15 13:44:45 2018 +0100

    Refactor a bit fswatch_command

    Avoid computing both commands eagerly.

    Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

commit c046b954b5d0186f6f9ea8dc12a653f5d593f3b2
Merge: fe7968da 98045261
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Wed Aug 15 11:03:09 2018 +0100

    Merge branch 'master' into watch-mode

commit fe7968da9fd2ce413d8b894bef6b34df650d2b1c
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Wed Aug 15 10:59:27 2018 +0100

    Fixed compilation error.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

commit b6612b1c6d39040a6e2689806422cee2ff1636a5
Author: Pavel Senchanka <pavel.senchanka@gmail.com>
Date:   Wed Aug 15 10:17:04 2018 +0100

    Added a simple version of a polling loop.

    `dune build -w` now behaves in a similar way to Jenga polling mode,
    automatically rebuilding the project when changes are detected.

    Note that on platforms other than Linux, you need to have `fswatch`
    installed for watch mode to work.

    Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

kodek16 committed:
Fix review suggestions.

Signed-off-by: Pavel Senchanka <pavel.senchanka@gmail.com>

kodek16 commented:
I've squashed the previous commits (it's nice that Github kept the review history reasonably clean, I wasn't sure it can do this), and fixed all of the suggestions above. @emillon Adding a newline to the message seems to preserve previous lines of Success/Failure, does it look better now?

emillon reviewed (approved):
The new feedback on build seems good to me. If you can amend the commit message so that it's only the "Add a simple version of the polling loop" it'd be perfect otherwise the concatenated messages are probably good enough :+1:

rgrinberg commented:
@kodek16 give this another rebase and we'll merge after CI.

jeremiedimino committed:
Merge branch 'master' into watch-mode

diml commented:
I did "Update branch" followed by "Squash and merge"
